src/data/fintype/card_embedding.lean

```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import data.fintype.card
+import data.nat.factorial
+
+/-!
+# Birthday Problem
+
+This file establishes the cardinality of `α ↪ β` in full generality.
+-/
+
+open finset
+
+local notation `|` x `|` := finset.card x
+local notation `‖` x `‖` := fintype.card x
+
+namespace fintype
+
+private lemma card_embedding_aux (n : ℕ) (β) [fintype β] [decidable_eq β] (h : n ≤ ‖β‖) :
+  ‖fin n ↪ β‖ = nat.desc_fac (‖β‖ - n) n :=
```

Comment:
I'm wondering if this would work more easily by the following argument:
* `X ⊕ Y ↪ Z` is equivalent to pairs of embeddings `X ↪ Z` and `Y ↪ Z` with disjoint images.
* an embedding with image disjoint from some set is equivalent to an embedding to the complement of that set (perhaps prove the function version first, or find it?)
* a calculation of the cardinality of a complement (hopefully already available)

It would certainly be more modular / reusable.

Reply:
(Oh, and then to do the induction step have by replacing `fin (n+1)` with `fin n ⊕ unit`.)

Reply:
Scott, I tried some of this in [here](https://github.com/leanprover-community/mathlib/commit/2a87ca3b791da24c05b4dfe20d8be07410b9ae51); would appreciate some comments on whether `function_thing` and `to_be_named` should be split up more, or everything should be defined inline.

as a bonus, though, this refactor seems to make the proof a looot shorter

Reply:
`function_thing` already exists as `sum_arrow_equiv_prod_arrow` (I found this by typing `⊕.*→.*≃` in the search box in VSCode). `to_be_named` looks good.

Reply:
I suspect that splitting `to_be_named` according to the first two bullet points I suggested above will be even easier (as then you can use `function_thing` for the "data" part of the equivalence. But perhaps not; try both ways?

Reply:
I tried a little splitting that up, but I wasn't sure how to express the disjointness too cleanly except as a subtype with disjoint set.ranges. I'll give it another try though!

Reply:
A more symmetric formulation of `to_be_named` might be `((α ⊕ β) ↪γ) ≃ (Σ s : set γ, α ↪ s × β↪ s.compl)`

Reply:
I'll put that on as a corollary, as I think it's probably of more general use, but it won't work in this case as it doesn't give me the size of `s`

Reply:
Alright, the new version is now up. It's not done yet (naming, docstrings, Eric's corollary) but the idea behind it is now clear; I have some questions in the new comments

Reply:
> I tried a little splitting that up, but I wasn't sure how to express the disjointness too cleanly except as a subtype with disjoint set.ranges. I'll give it another try though!

Did this not work? The `disjoint (set.range f) (set.range g)` version seems nicest to prove first. But I haven't actually tried it.

Reply:
just did it now scott; was scared of it and ended up being alright!